File: graham-stock-board/update_portfolio.py
import json, pathlib
from datetime import datetime, timezone

base = pathlib.Path(r'C:\Users\IkeFl\.openclaw\workspace\graham-stock-board')
portfolio_path = base / 'paper-trades.json'
portfolio = json.loads(portfolio_path.read_text(encoding='utf-8', errors='replace'))

quotes = {
  'PLTR': {'price': 148.35, 'chgPct': 0.2839180693571199, 'dayHigh': 149.46, 'dayLow': 144.45, 'prevClose': 147.93},
  'TEM': {'price': 46.40, 'chgPct': -1.9027484143763187, 'dayHigh': 47.16, 'dayLow': 45.32, 'prevClose': 47.30},
  'RKLB': {'price': 64.67, 'chgPct': -4.433279148810403, 'dayHigh': 68.30, 'dayLow': 64.12, 'prevClose': 67.67},
  'IONQ': {'price': 28.12, 'chgPct': -3.8303693570451354, 'dayHigh': 28.88, 'dayLow': 27.29, 'prevClose': 29.24},
  'SERV': {'price': 7.92, 'chgPct': -5.149700598802392, 'dayHigh': 8.29, 'dayLow': 7.80, 'prevClose': 8.35},
  'RXRX': {'price': 3.165, 'chgPct': 0.47619047619048016, 'dayHigh': 3.20, 'dayLow': 3.04, 'prevClose': 3.15},
}

for ticker, q in quotes.items():
    if ticker in portfolio.get('watchlist', {}):
        portfolio['watchlist'][ticker].update(q)

note = (
    '4/7/26 1:50 PM CT - All cash. No positions. '
    'PLTR $148.35 holding above $148 but still needs cleaner $150.50-151 confirmation. '
    'IONQ $28.12 still below $30 re-entry confirmation. '
    'TEM, RXRX, and SERV remain watch-only. '
    'No alerts fired. Starter zones remain valid. No trades made.'
)

portfolio['lastUpdated'] = '2026-04-07T18:50:00Z'
portfolio['note'] = note
portfolio['grahamNote'] = note
portfolio['strategy']['currentAssessment'] = note
portfolio['deployed'] = round(sum(float(p.get('cost', 0)) for p in portfolio.get('positions', [])), 2)
portfolio['deployedPct'] = round((portfolio['deployed'] / float(portfolio.get('portfolioSize', 0)) * 100) if portfolio.get('portfolioSize') else 0, 2)
portfolio['portfolioValue'] = round(float(portfolio.get('cash', 0)) + sum(float(p.get('shares', 0)) * float(p.get('currentPrice', 0)) for p in portfolio.get('positions', [])), 2)
portfolio['totalPnl'] = round(portfolio['portfolioValue'] - float(portfolio.get('portfolioSize', 0)), 2)
portfolio['totalPnlPct'] = round((portfolio['totalPnl'] / float(portfolio.get('portfolioSize', 0)) * 100) if portfolio.get('portfolioSize') else 0, 2)

portfolio_path.write_text(json.dumps(portfolio, indent=2), encoding='utf-8')
print('updated portfolio')

File: graham-stock-board/update-portfolio.js
const fs = require('fs');
const path = require('path');

const portfolioFile = path.join(__dirname, 'paper-trades.json');
const portfolio = JSON.parse(fs.readFileSync(portfolioFile, 'utf8'));

// Current prices from fetch
const prices = {
  PLTR: { price: 151.075, chgPct: -2.51 },
  TEM:  { price: 46.35,   chgPct: -0.88 },
  RKLB: { price: 69.00,   chgPct: -5.32 },
  IONQ: { price: 31.11,   chgPct: -2.66 }
};

const now = new Date().toISOString();
const timeLabel = "11:06 AM CT 3/26/26";

// Update open positions
portfolio.positions.forEach(pos => {
  const q = prices[pos.ticker];
  if (!q) return;
  pos.currentPrice = q.price;
  pos.marketValue = parseFloat((pos.shares * q.price).toFixed(2));
  pos.unrealizedPnl = parseFloat((pos.marketValue - pos.totalCost).toFixed(2));
  pos.unrealizedPnlPct = parseFloat(((pos.unrealizedPnl / pos.totalCost) * 100).toFixed(2));
  pos.note = `${timeLabel}. $${q.price} (${q.chgPct}% day). Holding ${pos.shares} shares, avg entry $${pos.avgEntry}. ${pos.unrealizedPnl >= 0 ? '+' : ''}$${pos.unrealizedPnl} unrealized (${pos.unrealizedPnlPct >= 0 ? '+' : ''}${pos.unrealizedPnlPct}%). Stop $${pos.stopLoss} intact — ${(((q.price - pos.stopLoss)/q.price)*100).toFixed(2)}% away. Zone $148-153 holding. Tape softening — RKLB ${prices.RKLB.chgPct}% IONQ ${prices.IONQ.chgPct}%. Next target: $${pos.target}.`;
});

// Recalculate deployed / portfolio value
const totalDeployed = portfolio.positions.reduce((s, p) => s + p.marketValue, 0);
portfolio.deployed = parseFloat(totalDeployed.toFixed(2));
portfolio.deployedPct = parseFloat(((totalDeployed / portfolio.portfolioSize) * 100).toFixed(2));
portfolio.portfolioValue = parseFloat((portfolio.cash + totalDeployed).toFixed(2));
portfolio.totalPnl = parseFloat((portfolio.portfolioValue - portfolio.portfolioSize).toFixed(2));
portfolio.totalPnlPct = parseFloat(((portfolio.totalPnl / portfolio.portfolioSize) * 100).toFixed(2));
portfolio.lastUpdated = now;

// Update watchlist
const pltrPos = portfolio.positions.find(p => p.ticker === 'PLTR');
portfolio.watchlist.PLTR = {
  price: prices.PLTR.price,
  chgPct: prices.PLTR.chgPct,
  note: `${timeLabel}. $${prices.PLTR.price} (${prices.PLTR.chgPct}%) — IN POSITION (4 shares). ${pltrPos ? (pltrPos.unrealizedPnl >= 0 ? '+' : '') + '$' + pltrPos.unrealizedPnl + ' unrealized (' + (pltrPos.unrealizedPnlPct >= 0 ? '+' : '') + pltrPos.unrealizedPnlPct + '%)' : ''}. Avg entry $149.79. Stop $144 (${(((prices.PLTR.price - 144)/prices.PLTR.price)*100).toFixed(2)}% away). Tape soft but zone holding. Next: $153-155 breakout add.`
};
portfolio.watchlist.TEM = {
  price: prices.TEM.price,
  chgPct: prices.TEM.chgPct,
  note: `${timeLabel}. $${prices.TEM.price} (${prices.TEM.chgPct}%) — still strongest name on board. PLTR ${prices.PLTR.chgPct}%, RKLB ${prices.RKLB.chgPct}%, IONQ ${prices.IONQ.chgPct}%. TEM holding relative strength. Zone $43-49. Best candidate for next buy when tape stabilizes.`
};
portfolio.watchlist.RKLB = {
  price: prices.RKLB.price,
  chgPct: prices.RKLB.chgPct,
  note: `${timeLabel}. $${prices.RKLB.price} (${prices.RKLB.chgPct}%) — $2.00 above $67 alert level. Re-entry zone $65-67. Tape still weak. Not entering without clear reversal candle and volume.`
};
portfolio.watchlist.IONQ = {
  price: prices.IONQ.price,
  chgPct: prices.IONQ.chgPct,
  note: `${timeLabel}. $${prices.IONQ.price} (${prices.IONQ.chgPct}%). Re-entry zone $28-29, still $2.11 above zone. No action. Stopped out at $31.00 — not chasing back in.`
};

// Update strategy assessment
portfolio.strategy.currentAssessment = `${timeLabel}. HOLD — No trades. PLTR $${prices.PLTR.price} (${prices.PLTR.chgPct}%), ${pltrPos ? (pltrPos.unrealizedPnl >= 0 ? '+' : '') + '$' + pltrPos.unrealizedPnl + ' unrealized (' + (pltrPos.unrealizedPnlPct >= 0 ? '+' : '') + pltrPos.unrealizedPnlPct + '%)' : ''} on 4 shares. TEM $${prices.TEM.price} (${prices.TEM.chgPct}%) still strongest name on board. RKLB $${prices.RKLB.price} (${prices.RKLB.chgPct}%) — $2.00 above re-entry zone. IONQ $${prices.IONQ.price} (${prices.IONQ.chgPct}%) — $2.11 above re-entry. ${portfolio.deployedPct}% deployed, ${(100-portfolio.deployedPct).toFixed(2)}% cash. Capital well-protected.`;

portfolio.grahamNote = `${timeLabel}. HOLD. PLTR $${prices.PLTR.price} (${prices.PLTR.chgPct}%) — 4 shares, avg entry $149.79. ${pltrPos ? (pltrPos.unrealizedPnl >= 0 ? '+' : '') + '$' + pltrPos.unrealizedPnl + ' unrealized' : ''}. Stop $144 still ${(((prices.PLTR.price - 144)/prices.PLTR.price)*100).toFixed(2)}% away, well intact. TEM $${prices.TEM.price} (${prices.TEM.chgPct}%) continues to outperform — best relative-strength name. RKLB at $${prices.RKLB.price} (${prices.RKLB.chgPct}%) still $2 above my $67 re-entry line — not touching it. IONQ $${prices.IONQ.price} (${prices.IONQ.chgPct}%) needs to come back to $29 before I'd reconsider. Broad tape soft, no urgency to deploy. ${portfolio.deployedPct}% deployed. Watching.`;

fs.writeFileSync(portfolioFile, JSON.stringify(portfolio, null, 2));
console.log(JSON.stringify({
  status: 'updated',
  pltr: { price: prices.PLTR.price, chgPct: prices.PLTR.chgPct, unrealizedPnl: pltrPos ? pltrPos.unrealizedPnl : null },
  portfolioValue: portfolio.portfolioValue,
  totalPnl: portfolio.totalPnl,
  totalPnlPct: portfolio.totalPnlPct
}, null, 2));
